supportedBackends/aws/src/main/resources/sidecar/sidecar

```diff
@@ -0,0 +1,117 @@
+      echo Target file found - killing inotifywait and returning
+      killall inotifywait # inotifywait will simply hang out with a return like this
+      return
+    fi
+    relativePath=${path#"${localCallRoot}/"}
+    localDiskRelativePath=${path#"${AWS_CROMWELL_LOCAL_DISK}/"}
+    if [ "$action" = "MOVED_TO" ] || [ "$action" = "CLOSE_WRITE,CLOSE" ]; then
+      case $path in
+        ${AWS_CROMWELL_LOCAL_DISK}/${exclude}*)
+          echo skipping change in excluded directory
+          ;;
+        *)
+          aws s3 cp "${path}${file}" "$AWS_CROMWELL_CALL_ROOT/${localDiskRelativePath}${file}"
+      esac
+    fi
+    if [ "$action" = "MOVED_FROM" ]; then
+      case $path in
+        ${AWS_CROMWELL_LOCAL_DISK}/${exclude}*)
+          echo skipping change in excluded directory
+          ;;
+        *)
+          aws s3 rm "$AWS_CROMWELL_CALL_ROOT/${localDiskRelativePath}${file}"
+      esac
+    fi
+  done
+}
+
+###########################################################################
+# The general plan is to do a sync of all workflow files into the container
+# We share the same volumes/mount points of the original, so we can go to
+# town here. Generally, we assume all *inputs* come from
+# AWS_CROMWELL_WORKFLOW_ROOT
+# All *outputs* will eventually end up in AWS_CROMWELL_CALL_ROOT.
+# However, within the container, this root is effectively AWS_CROMWELL_LOCAL_DISK
+#
+# The trick here is that the inputs are also landed inside
+# AWS_CROMWELL_LOCAL_DISK. So we need to exclude the inputs from being
+# synced (both real time and in batch at the end), so we do some fancy
+# stuff for that
+###########################################################################
+# call root example: s3://cromwell-aws/cromwell-execution/...
+# localRoot example: /cromwell_root/cromwell-aws/cromwell-execution...
+# AWS_CROMWELL_PATH example: ...
+localCallRoot="$AWS_CROMWELL_LOCAL_DISK"/"${AWS_CROMWELL_CALL_ROOT#"s3://"}"
+localWorkflowRoot="$AWS_CROMWELL_LOCAL_DISK"/"${AWS_CROMWELL_WORKFLOW_ROOT#"s3://"}"
+suffix="/*"
+exclude=${AWS_CROMWELL_CALL_ROOT#"s3://"}
+exclude=${exclude%%$suffix}
+echo "syncing files from s3"
+echo "source ${AWS_CROMWELL_WORKFLOW_ROOT}, dest $localWorkflowRoot"
+aws s3 sync "$AWS_CROMWELL_WORKFLOW_ROOT" "$localWorkflowRoot"
+echo "Resuming container"
+docker unpause "$AWS_CROMWELL_TASK_DOCKER_ID"
+# Place watch for the rc file
+waitfor "$AWS_CROMWELL_LOCAL_DISK" "${AWS_CROMWELL_RC_FILE#"${AWS_CROMWELL_LOCAL_DISK}/"}"
+
+# We'll pause the task container while we do our work. This will prevent
+# the container from exiting before files are out on S3. Also, the ecs agent
+# has some cleanup, and we don't want a race condition with that process
+docker pause "$AWS_CROMWELL_TASK_DOCKER_ID"
+
+###########################################################################
+# Now we will do the final sync. If we've done the job right, this sync
+# should only copy the rc file in place to s3, since all the other files
```

Comment:
Why do we need to re-sync everything if we know that the only thing left is the rc file ?
In what circumstance could we have not "done the job right" ?

Reply:
In my testing, all files are synced properly by the time this call is run, so it's effectively a no-op, with the exception of the rc file which I'm currently not copying prior to the exit from waitfor(). There is, however, a possibility of missed files due to issues with inotifywait that are documented in the man page: https://linux.die.net/man/1/inotifywait

Reply:
To take the problem the other way around then, why not _only_ do a sync of everything at the end instead of copying files individually with `inotifywait` if it has issues ? Just trying to see if the redundancy is really needed :)

Reply:
That was the original design, but I was told I should be doing a sync-as-you-go. I'm happy to pull out the inotify stuff and just to the pre and post sync steps though if that's all that's needed. It would have the advantage of not syncing temp files.

Reply:
No - the files being processed by the current running task.

Reply:
That's correct - I didn't want to sync to S3 during file writes, so I'm doing it after close. The theory being that in a days-long workflow, files written would be available even if the full processed dataset is not.